### PR TITLE
Validate private data

### PR DESCRIPTION
**Pull request by phoebe-lew**

### CODE REVIEW

File: tbdex/rfq/rfq.go [hunk 1]
```diff
@@ -60,18 +61,10 @@ func (r *RFQ) UnmarshalJSON(data []byte) error {
 	// }
 
 	// TODO verify private data
```
Comment (diehuxx) [inline tbdex/rfq/rfq.go]: remove todo?

File: tbdex/rfq/rfq.go [hunk 2]
```diff
@@ -34,8 +35,8 @@ func (r *RFQ) Sign(bearerDID did.BearerDID) error {
 	return nil
 }
 
-// UnmarshalJSON validates and unmarshals the input data into an RFQ.
-func (r *RFQ) UnmarshalJSON(data []byte) error {
+// ValidateAndUnmarshalJSON validates and unmarshals the input data into an RFQ.
+func (r *RFQ) ValidateAndUnmarshalJSON(data []byte, verifyPrivateDataStrict bool) error {
```
Comment (diehuxx) [inline tbdex/rfq/rfq.go]: low key i prefer `UnmarshallJSON`. i asked chad and he thinks it's okay to have custom validations in an `UnmarshallJSON` method.

File: tbdex/rfq/rfq_test.go [hunk 3]
```diff
@@ -81,7 +81,57 @@ func TestRFQ_Sign(t *testing.T) {
...
+
+	bytes, err := json.Marshal(r)
+	assert.NoError(t, err)
+
+	var rfq rfq.RFQ
+	err = rfq.ValidateAndUnmarshalJSON(bytes, false)
+	assert.NoError(t, err)
+}
+
+func TestRFQ_UnmarshalJSON_FailsIncompletePrivateData(t *testing.T) {
```
Comment (diehuxx) [inline tbdex/rfq/rfq_test.go]: I'm not sure why unmarshalling should fail in this test. `rfq.Create` should output the RFQ with all the private data, no? why are is marshalling producing an error?

File: tbdex/rfq/rfq.go [hunk 4]
```diff
@@ -200,4 +193,37 @@ func (r RFQ) Digest() ([]byte, error) {
...
+	}
+
+	if strict || r.Data.ClaimsHash != "" {
+		payload := []any{r.PrivateData.Salt, r.PrivateData.Claims}
+		if err := tbdex.VerifyDigest(r.Data.ClaimsHash, payload); err != nil {
+			return fmt.Errorf("failed to verify claims: %w", err)
+		}
+	}
+
+	if strict || r.Data.Payin.PaymentDetailsHash != "" {
```
Comment (diehuxx) [inline tbdex/rfq/rfq.go]: if `PaymentDetailsHash` is null, then we shouldn't validate. that means that the payin has no payment details, which is still valid for RFQs.

File: tbdex/rfq/rfq.go [hunk 5]
```diff
@@ -200,4 +193,37 @@ func (r RFQ) Digest() ([]byte, error) {
...
+		payload := []any{r.PrivateData.Salt, r.PrivateData.Payin.PaymentDetails}
+		if err := tbdex.VerifyDigest(r.Data.Payin.PaymentDetailsHash, payload); err != nil {
+			return fmt.Errorf("failed to verify payin: %w", err)
+		}
+	}
+
+	if strict || r.Data.Payout.PaymentDetailsHash != "" {
```
Comment (diehuxx) [inline tbdex/rfq/rfq.go]: ditto
https://github.com/TBD54566975/tbdex-go/pull/20/files#r1563138515

File: tbdex/rfq/rfq.go [hunk 6]
```diff
@@ -200,4 +193,37 @@ func (r RFQ) Digest() ([]byte, error) {
...
+func (r *RFQ) verifyPrivateData(strict bool) error {
+	if r.PrivateData == nil {
+		if strict {
+			return errors.New("private data is missing")
+		}
+
+		return nil // no PrivateData to verify
+	}
+
+	if strict || r.Data.ClaimsHash != "" {
```
Comment (diehuxx) [inline tbdex/rfq/rfq.go]: ditto for `claimsHash`. if `claimsHash` is null, then we shouldn't perform the validation.
https://github.com/TBD54566975/tbdex-go/pull/20/files#r1563138515

File: tbdex/rfq/rfq.go (hunk 2 above)
Comment (mistermoe) [reply tbdex/rfq/rfq.go]: Agreed minus the additional `l` in `Unmarshal` bc go interface

File: tbdex/rfq/rfq_test.go
Comment (diehuxx) [inline tbdex/rfq/rfq_test.go:1]: Since the private data stuff has a lot of edge cases, can we add more tests for both success and failure cases for both strict and nonstrict for each of payin, payout, and claims?

Comment (diehuxx) [reply tbdex/rfq/rfq_test.go:1]: Even for me who has implemented this three times, it's a dense chunk of code to comb through.

**Comment by phoebe-lew** [comment]
@diehuxx not ready for review

File: tbdex/rfq/rfq.go (hunk 2 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq.go]: @mistermoe what's the suggestion here? `UnmarshalJSON`? 

Changed it because I was getting an error that a method named `UnmarshalJSON` should have the signature `UnmarshalJSON(data []byte)` without the bool.

File: tbdex/rfq/rfq_test.go (hunk 3 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq_test.go]: marshaling is not producing an error, UNmarshaling is. This is testing that unmarshal fails if `strict` is true, which is how I interpreted the method should work based on the docs in `tbdex-js`:

```
/**
   * Verify the presence and integrity of all possible properties in {@link Rfq.privateData}.
   * @throws if there are properties missing in {@link Rfq.privateData} or which do not match the corresponding
   *         hashed property in {@link Rfq.data}
   */
```

Let me know if it should function differently!

File: tbdex/vectors_test.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+// 	input := offering.Offering{}
+// 	input.UnmarshalJSON([]byte(vector.Input))
+
+// 	assert.Equal(t, input, vector.Output)
+// }
+
+func TestRFQVectors(t *testing.T) {
+	vector := readVector[rfqVector]("parse-rfq.json")
+	input := rfq.RFQ{}
+	err := input.ValidateAndUnmarshalJSON([]byte(vector.Input), false)
```
Comment (phoebe-lew) [inline tbdex/vectors_test.go]: @mistermoe failing with:

```
=== RUN   TestRFQVectors
    /Users/plew/Development/tbdex-go/tbdex/vectors_test.go:62: Did not expect an error but got:
        failed to verify RFQ signature: kid does not match any verification method no verification method found for id: did:dht:9kkuh34q7nkd4tphbcg7py9h1g16iftbtskesi9courdwj96q3sy#0
```

which is coming from https://github.com/TBD54566975/web5-go/blob/main/jws/jws.go#L247-L250

does that seem expected?

File: tbdex/rfq/rfq.go (hunk 2 above)
Comment (mistermoe) [reply tbdex/rfq/rfq.go]: Aye bc https://pkg.go.dev/encoding/json#Unmarshaler

We could scrap the bool and just verify `privateData` if it's not nil

File: tbdex/vectors_test.go (hunk 7 above)
Comment (mistermoe) [reply tbdex/vectors_test.go]: @phoebe-lew It doesn't seem expected! I just fixed a dht bug last night. Maybe bump `webfif-go` to 0.14.0 and try agane? 🤞🏽

Comment (phoebe-lew) [reply tbdex/vectors_test.go]: @mistermoe still failing, printed out the DID doc:

```
vmSelector: did:dht:9kkuh34q7nkd4tphbcg7py9h1g16iftbtskesi9courdwj96q3sy#0
resolutionResult.Document: {
  "id": "did:dht:9kkuh34q7nkd4tphbcg7py9h1g16iftbtskesi9courdwj96q3sy",
  "verificationMethod": [
    {
      "id": "0",
      "type": "JsonWebKey",
      "controller": "",
      "publicKeyJwk": {
        "kty": "OKP",
        "crv": "Ed25519",
        "x": "-pU-Z07olD1FvAsN1oP8kaXqliGNlItX7ITIOif-dmw"
      }
    }
  ],
  "assertionMethod": [
    "0"
  ],
  "authentication": [
    "0"
  ],
  "capabilityDelegation": [
    "0"
  ],
  "capabilityInvocation": [
    "0"
  ]
}
```

The `SelectVerificationMethod` should be looking for a match for `0`, correct?

Comment (mistermoe) [reply tbdex/vectors_test.go]: ah ok i see it. the IDs are borked in that doc. needs to be `#0`

Comment (phoebe-lew) [reply tbdex/vectors_test.go]: Would that be an issue in web5-go where resolution happens and returns a document or in the SDK where we created the DID (i'm assuming web5-js)?

File: tbdex/rfq/rfq.go (hunk 2 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq.go]: Yea that would fix the error, do you know why we needed the bool here? https://github.com/TBD54566975/tbdex-js/blob/main/packages/protocol/src/message-kinds/rfq.ts#L78-L82

File: tbdex/rfq/rfq_test.go (hunk 3 above)
Comment (diehuxx) [reply tbdex/rfq/rfq_test.go]: It's possible I'm misreading the code. I think that `rfq.Create(...)` returns a valid RFQ with the PrivateData. So marshalling + unmarshalling the RFQ without modifying the JSON should succeed. Are we detaching the private data somewhere before we call `rfq.ValidateAndUnmarshalJSON(bytes, true)`?

When `strict` is true, we should verify that the private data is expected present and correct. E.g. if `data.claimsHash != null`, then we should check that `privateData.claims` is present and correct. But if `data.claimsHash == null`, then we should NOT expect `privateData.claims` to be present because the RFQ contains no claims.

File: tbdex/vectors_test.go (hunk 7 above)
Comment (mistermoe) [reply tbdex/vectors_test.go]: taking a look rn

Comment (mistermoe) [reply tbdex/vectors_test.go]: https://github.com/TBD54566975/web5-go/pull/138

File: tbdex/rfq/rfq.go [hunk 8]
```diff
@@ -59,19 +65,86 @@ func (r *RFQ) UnmarshalJSON(data []byte) error {
...
+
+	hashed, err := tbdex.DigestJSON(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to digest rfq: %w", err)
+	}
+
+	return hashed, nil
+}
+
+func (r *RFQ) verifyPrivateData(strict bool) error {
```
Comment (phoebe-lew) [inline tbdex/rfq/rfq.go:103]: seeking feedback on this function, prioritized collapsing both strict/not strict into a single method but could split into 2 if we think the readability outweighs

Comment (diehuxx) [reply tbdex/rfq/rfq.go:103]: 3 levels of nested `if`s is kinda hard to read, but i won't make you split it into two methods if you prefer it this way. I preferred this pattern that I used in [tbdex-js](https://github.com/TBD54566975/tbdex-js/blob/main/packages/protocol/src/message-kinds/rfq.ts#L180-L256). Splitting out methods `verifyAllPrivateData` and `verifyPresentPrivateData` which maps roughly to `strict == true` and `strict == false`. Those two methods each call `verifyClaimsHash`, `verifyPayinHash` and `verifyPayoutHash`. This kept each method pretty focused -- no nested `if`s. Ultimately it's a matter of style, so I'll leave it up to you which feels better.

File: tbdex/rfq/rfq_test.go [hunk 9]
```diff
@@ -81,7 +81,117 @@ func TestRFQ_Sign(t *testing.T) {
...
+
+func TestRFQ_Unmarshal_Invalid(t *testing.T) {
+	input := []byte(`{"doo": "doo"}`)
+
+	var rfq rfq.RFQ
+	err := rfq.UnmarshalJSON(input)
+	assert.Error(t, err)
+}
+
+func TestRFQ_Verify_NoPrivateDataStrict(t *testing.T) {
```
Comment (diehuxx) [inline tbdex/rfq/rfq_test.go:113]: Confused by this test name. "no private data" sounds like the private data has been detached, which should cause `rfq.Verify()` to fail. is the private data being detached somehow?

File: tbdex/rfq/rfq_test.go [hunk 10]
```diff
@@ -81,7 +81,117 @@ func TestRFQ_Sign(t *testing.T) {
...
+
+	var rfq rfq.RFQ
+	err = rfq.UnmarshalJSON(bytes)
+	assert.NoError(t, err)
+
+	err = rfq.Verify(true)
+	assert.NoError(t, err)
+}
+
+func TestRFQ_Verify_NoPrivateDataNotStrict(t *testing.T) {
```
Comment (diehuxx) [inline tbdex/rfq/rfq_test.go:139]: similar to https://github.com/TBD54566975/tbdex-go/pull/20/files#r1569280959, the test name "no private data" makes me think the private data has been detached.

File: tbdex/rfq/rfq_test.go [hunk 11]
```diff
@@ -106,12 +216,156 @@ func TestRFQ_UnmarshalJSON(t *testing.T) {
...
+		rfq.Payin("100", "STORED_BALANCE"),
+		rfq.Payout("BANK_ACCOUNT", rfq.PaymentDetails(
+			map[string]any{
+				"accountNumber": "1234567890123456",
+				"routingNumber": "123456789",
+			})),
+	)
+
+	_ = r.Sign(walletDID)
+	r.PrivateData.Payout.PaymentDetails = nil
```
Comment (diehuxx) [inline tbdex/rfq/rfq_test.go:368]:
```suggestion
func TestRFQ_Verify_FailsMissingDataForPayinHashStrict(t *testing.T) {
	pfiDID, _ := didjwk.Create()
	walletDID, _ := didjwk.Create()
	offeringID, _ := typeid.WithPrefix(offering.Kind)

	r, _ := rfq.Create(
		walletDID.URI,
		pfiDID.URI,
		offeringID.String(),
		rfq.Payin("100", "STORED_BALANCE"),
		rfq.Payout("BANK_ACCOUNT", rfq.PaymentDetails(
			map[string]any{
				"accountNumber": "1234567890123456",
				"routingNumber": "123456789",
			})),
	)

	_ = r.Sign(walletDID)
	r.PrivateData.Payin.PaymentDetails = nil
```

File: tbdex/rfq/rfq_test.go [hunk 12]
```diff
@@ -81,7 +81,117 @@ func TestRFQ_Sign(t *testing.T) {
...
+
+	var rfq rfq.RFQ
+	err = rfq.UnmarshalJSON(bytes)
+	assert.NoError(t, err)
+
+	err = rfq.Verify(false)
+	assert.NoError(t, err)
+}
 
+func TestRFQ_Verify_FailsPrivateDataHashMismatch(t *testing.T) {
```
Comment (diehuxx) [inline tbdex/rfq/rfq_test.go]: do you mind making similar tests that check when payin and payout hash mismatch?

**Review by diehuxx — APPROVED** [review]
nice work. a couple small things, mostly about tests. otherwise lgtm

File: tbdex/rfq/rfq.go [hunk 13]
```diff
@@ -59,19 +65,86 @@ func (r *RFQ) UnmarshalJSON(data []byte) error {
...
+		return fmt.Errorf("failed to verify private data: %w", err)
+	}
 
-	// 	}
-	// }
+	return nil
+}
+
+// Parse validates, parses input data into an RFQ, and verifies the signature and private data.
+func (r *RFQ) Parse(data []byte, privateDataStrict bool) error {
```
Comment (jiyoonie9) [inline tbdex/rfq/rfq.go:77]: i would expect `Parse()` method to return an RFQ?

File: tbdex/rfq/rfq.go [hunk 14]
```diff
@@ -59,19 +65,86 @@ func (r *RFQ) UnmarshalJSON(data []byte) error {
...
+	hashed, err := tbdex.DigestJSON(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to digest rfq: %w", err)
+	}
+
+	return hashed, nil
+}
+
+func (r *RFQ) verifyPrivateData(strict bool) error {
+	if strict == false {
```
Comment (jiyoonie9) [inline tbdex/rfq/rfq.go]: nit: collapse into 1 if? `if (strict == false && r.PrivateData == nil) { return nil }`

File: tbdex/rfq/rfq_test.go [hunk 15]
```diff
@@ -81,7 +81,117 @@ func TestRFQ_Sign(t *testing.T) {
...
+
+	r, _ := rfq.Create(
+		walletDID.URI,
+		pfiDID.URI,
+		offeringID.String(),
+		rfq.Payin("100", "STORED_BALANCE"),
+		rfq.Payout("BANK_ACCOUNT"),
+	)
+
+	_ = r.Sign(walletDID)
```
Comment (jiyoonie9) [inline tbdex/rfq/rfq_test.go:95]: do we need the `_ =` in order for linter to not complain?

**Review by jiyoonie9 — APPROVED** [review]

File: tbdex/rfq/rfq_test.go (hunk 9 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq_test.go:113]: There's no private data in the RFQ at all.

File: tbdex/rfq/rfq_test.go (hunk 11 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq_test.go:368]: PaymentDetails is on payout?

```
rfq.Payout("BANK_ACCOUNT", rfq.PaymentDetails(
			map[string]any{
				"accountNumber": "1234567890123456",
				"routingNumber": "123456789",
			})),
```

File: tbdex/rfq/rfq_test.go (hunk 15 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq_test.go:95]: sign returns an error, but i'm not checking it as that functionality was tested above.

File: tbdex/rfq/rfq.go (hunk 13 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq.go:77]: Following the pattern of how `json.Unmarshal` works, Parse is a receiver method so it is modifying `r`. Usage:

```
res = rfq.RFQ{}
err = res.Parse([]byte(vector.Input), false)
```

Similar to the json pattern:

```
ret := rfq{}
err = json.Unmarshal(data, &ret)
if err != nil {
	return fmt.Errorf("failed to JSON unmarshal rfq: %w", err)
}
```

File: tbdex/rfq/rfq_test.go (hunk 9 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq_test.go:113]: It's checking for `for strict = true, if all the hashes are null, then is private data allowed to be null`

File: tbdex/rfq/rfq_test.go (hunk 10 above)
Comment (phoebe-lew) [reply tbdex/rfq/rfq_test.go:139]: This is the test case for `strict = false, private data is nil, then return without error`. Do you have a suggestion for a better name?
